man/dbvform.1
.TH DBVFORM 1 "12 September 1995"
.FB
.SH NAME
dbvform \- format a \fIFirstBase\fP database using standard view dictionary
.SH SYNOPSIS
.B dbvform
[\fB-d\fP \fIdatabase\fP] [\fB-i\fP \fIindex\fP] [\fB-v\fP \fIview\fP]
[\fB-f\fP] [\fB-B\fP]
[\fB-l\fP \fIpage-length\fP] [\fB-w\fP \fIpage-width\fP]
.SH DESCRIPTION
.I Dbvform
formats undeleted records of a \fIFirstBase\fP database using a standard
view dictionary as described in \fIview\fP(5). All output is written to 
standard output, usually the terminal.
.PP
A logical page is generated for each record of the database or index,
as well as for each page within the view dictionary, as denoted by the
$PAGE lines (\fIview\fP(5)).
.PP
Without an index, \fIdbvform\fP(1) will process the entire database.
Using an index, a very select portion of a database can be formatted.
.PP
The \fB-B\fP flag will enable the fastest throughput of
records from the database. If this flag can be used, the \fIdbvform\fP
generation is speeded up a great deal by internal record buffering.
Note that no index can be used with record buffering, and
that record locking is disabled.
.PP
The length and width of the output can be specified using the \fB-l\fP
and \fB-w\fP flags \(em with the default being 24 lines by 80 columns, and
the maximum being 66 lines by 132 columns.
.PP
Adjusting the page lengths and widths can create some very desirable
effects. For example, a page length of \fB1\fP (one) with a proper view
dictionary that maps a few fields to line one will create a tabular listing
of the specified data. Or, using a length and width of \fB9\fP
and \fB40\fP, badges or labels could be generated.
.PP
\fIDbvform\fP will normally generate \fIpage-length\fP lines for each
record processed, even if the lines
are empty. Using the \fB-f\fP flag will cause formfeeds to be used instead.
.SH FILES
.PD 0
.TP 10
dbase
default name of database.
.TP 10
dbase
default name of view dictionary.
.TP 10
index
default name of index.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.map
record map of \fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary file.
.PD
.SH SEE ALSO
dbvedit(1), view(5)
.PP
.I FirstBase User's Guide and Reference Manual
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbvform/dbvform.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbvform.c,v 9.0 2001/01/09 02:56:03 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbvform.c	8.2 04 Aug 1997 FB";
#endif

/*  
 *  dbvform.c - vform records using a standard vdict (view(5)) file.
 */

#include <dbvf_v.h>

static char *USAGE = 
"usage: dbvform [-w width] [-l length] [-f] [-d dbase] [-i index] [-v view]";

static Bflag = 0;				/* Blocking flag */
static sform();
static onenode();

extern short int cdb_returnerror;

/*
 *  dbvform - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
   
      {
	 int j;

	 cdb_batchmode = 1;
	 cdb_returnerror = 1;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 p_cols = cdb_t_cols;
	 p_lines = cdb_t_lines;
	 if (fb_testargs(argc, argv, "-f") > 0)
	    formfeed_flag = 1;
	 if (fb_testargs(argc, argv, "-B") > 0)
	    Bflag = 1;
	 if ((j = fb_testargs(argc, argv, "-w")) > 0)
	    p_cols = atoi(argv[j + 1]);
	 if ((j = fb_testargs(argc, argv, "-l")) > 0)
	    p_lines = atoi(argv[j + 1]);
	 if (p_cols <= 0)
	    p_cols = cdb_t_cols;
	 if (p_lines <= 0)
	    p_lines = cdb_t_lines;
	 if (p_cols <= 0)
	    p_cols = 80;
	 if (p_lines <= 0)
	    p_lines = 24;
	 if (p_lines > F_PROW || p_cols > F_PCOL)
	    usage();
	 F_fb_initscreen();
	 fb_allow_int();		/* allows instant piping */	
	 vform(argc, argv);
	 fb_ender();
      }

/*
 *  vform - emit loop for a single dbase.
 */

   vform(argc, argv)
      int argc;
      char *argv[];
   
      
      {
         int sform(), st;

	 if (fb_opendb(cdb_db, READ, FB_WITHINDEX, FB_OPTIONAL_INDEX)!=FB_AOK){
	    fb_serror(FB_MESSAGE, SYSMSG[S_BAD_DATA], cdb_db->dbase);
	    return(FB_ERROR);
	    }
	 if (fb_get_vdict(argc, argv) == FB_ERROR)
	    fb_xerror(FB_BAD_DICT, vdict, NIL);
	 if (cdb_db->dindex == NULL || cdb_db->dindex[0] == NULL ||
               cdb_db->ifd < 0){
            if (Bflag == 0)
	       st = fb_foreach(cdb_db, sform);
            else
	       st = fb_blockeach(cdb_db, sform);
            }
	 else
	    st = fb_forxeach(cdb_db, sform);
	 fb_closedb(cdb_db);
	 return(st);
      }

/* 
 *  sform - sform a single record.
 */

   static sform(db)
      fb_database *db;
      
      {
         fb_page *p;
	 fb_node *n;
	 
	 for (p = phead; p; p = p->p_next){
	    for (n = p->p_nhead; n; n = n->n_next)
	       onenode(n);
	    screendump();
	    }
         return(FB_AOK);
      }

/*
 * onenode - vform a single node
 */

   static onenode(n)
      fb_node *n;
      
      {
         fb_field *f;

	 F_fb_move(n->n_row, n->n_col);
	 if (n->n_fp == NULL){		/* i.e., if its not a fb_field */
	    F_fb_printw(FB_FSTRING, n->n_text);
	    }
	 else{
	    f = n->n_fp;
	    fb_fetch(f, cdb_afld, cdb_db);
	    frm_putfield(n, f, cdb_afld);
	    }
      }

/* 
 * usage message 
 */

   usage()
      {
         fb_xerror(FB_MESSAGE, USAGE, NIL);
      }
